### ISSUE TITLE
[Feature] Federation between gogs instances

### ISSUE THREAD

**Issue by yookoala** [gogits/gogs#4437]
## Description

The elevator pitch would be this:

> "It would be great to fork / send pull request between public repositories across different gogs installation. So it would be great to have a way to federate between gogs instances."

My suggestion would be to look at [ActivityPub](https://www.w3.org/TR/activitypub/), in-particular the "Federation Protocol", and see if we can federate through that protocol. The protocol is for building decentralized social network. But I think it is extensible with custom message `type` like "`Fork`" or "`Pull-request`".

**Comment by bachy** [comment]
ho yes please please please (begging dog face)

**Comment by unlmtd** [comment]
This is the github-killer feature.

**Comment by wakawakaeldondolero** [comment]
Yes please add federation! For me this is the n1 priority! Definitely the github killer feature!

**Comment by labdsf** [comment]
ActivityPub is for stars (called likes in ActivityPub) and following.

For forks, there is https://webmention.net/ This protocol can be used to notify source repository whenever it is cloned via gogs mirror/migration feature. It is indented for mentioning in blog posts, but can be easily adopted for forks and pull requests.

**Comment by yookoala** [comment]
We may consider "Fork" as a custom "ActivityType" as described by [ActivityStreams Vocabs](https://www.w3.org/TR/activitystreams-vocabulary/), an ActivityStreams extension.

**Comment by Skehmatics** [comment]
Another project that may be of use: https://github.com/go-fed/activity

**Comment by yookoala** [comment]
Relevant proposal in gitea repository:
https://github.com/go-gitea/gitea/issues/1612

**Comment by yookoala** [comment]
Let's lay down (and discuss) what needed to be done first.

### Terms
**Git Service**: Web based Git service like Gogs, Gitea, Gitlab.
**Web Repository**: Web based entity on a Git Service that represents an underlying Git repository branch.
**upstream**: A git repository where a PR is being sent TO.
**downstream**: A git repository where a PR is being sent FROM.

### Features

For federation, these are the base line features:
1. Allow user to create a Pull Request to a remote upstream Web Repository.
2. Allow a downstream Web Repository to receive a Pull Request from a downstream Web Repository.
3. Allow a downstream Web Repository to report updates (e.g. git push) of a branch to an upstream Web Repository that it previously sent Pull Request to.
4. Allow an upstream Web Repository to notify a remote Git Service that a Pull Request is commented, tagged, assigned, reviewed, closed, re-openned or otherwise updated.

Based on these feature needs, It might also be essential to allow any Web Repository (specific branch) to be properly addressable and discover-able by query protocol (e.g. Like how Mastodon use [Web Finger](https://tools.ietf.org/html/rfc7033)). This universal address should be differ from user or organization.

**Comment by 21stio** [comment]
What about establishing an open protocol like ActivityHub in parallel to that?

**Comment by 21stio** [comment]
I secured gits.social and gogs.social, so if anybody is going to make it happen and is interested in the domains, let me know and I'll give them to you

**Comment by cwebber** [comment]
Hi!  ActivityPub co-editor here.  I'm fairly busy at the moment but I'd like to see this happen... if you need questions feel free to ping me, or ask in #social on irc.w3.org

**Comment by benyanke** [comment]
This would indeed be killer - my main concern with moving off github is the lack of "social interface" that you get.

If this can somehow replicate that to some degree....that'd be amazing.

**Comment by yookoala** [comment]
Some more thoughts: We'd also need to cater for Forking.

Downstream should notify the upstream about their forking address (so upstream could tell the owner). Potentially, a downstream would wants to know about upstream changes and vice verse. So a general subscription would be needed (which should comes natural to ActivityPub).

**Feature** (appending)

5. Allow user to create a Fork notification to a remote upstream Web Repository.
6. Allow an upstream Web Repository to receive a Fork notification from a downstream Web Repository (not branch specific).
7. Allow any agent to subscribe to any Web Repository updates.

**Comment by tobast** [comment]
An even more killer feature would be to make sure every major git tracking system uses the same standard, so that it would be able to federate Gogs with Gitlab, Gitea, … and why not even Github if the protocol gets big enough.

**Comment by yookoala** [comment]
@cwebber: I plan to draft a specification document in this repository:
https://github.com/git-federation/gitpub

I feel like it is the best move to kick this off. But I'm have no experience in drafting such a thing. Please advice / join the effort. Thanks.

**Comment by jaywink** [comment]
> An even more killer feature would be to make sure every major git tracking system uses the same standard, so that it would be able to federate Gogs with Gitlab, Gitea, 

I think this should even be the most important thing to consider before implementing anything - getting some specs agreed so that platforms can actually talk to each other.

Good stuff! <3

**Comment by 21stio** [comment]
just to be clear, the look and feel would be something like if I have a private gog instance and I search for a repo, and it's hosted on e.g. a gitlab instance, I would see that repo in my gog ui, right? 

and the clone address would still point to the gitlab instance or?

**Comment by tobast** [comment]
> just to be clear, the look and feel would be something like if I have a private gog instance and I search for a repo, and it's hosted on e.g. a gitlab instance, I would see that repo, issues etc. in my gog ui, right?

This looks a bit overkill to me. The original idea, as I understand it (please correct me if I'm wrong) would be (at least in a first place) to federate PRs and forks, to make them behave "as if" the two repos were on the same gogs/gitlab/whatever instance.

Eg., I would be able to fork the gitlab project (hosted on gitlab) on my own Gogs instance, and this fork would be referenced there, so that the original project can track public forks. Then, I would be able to submit painlessly a PR to the original repo.

**Comment by 21stio** [comment]
hm.. to me federation is when a lot of people share n bigger instances that are interconnected with each other, so only a few people need to run a server and others somehow join

what you are talking about sounds a little like a maintained version of gittorrent, if Im not mistaken https://news.ycombinator.com/item?id=9625840
https://github.com/cjb/gittorrent

**Comment by yookoala** [comment]
@21stio: Federation does mean what you say. Gogs / Gitea / Gitlab instances are those n bigger instances. And the goal for confederation is to have their major workflow inter-connectable. So the specification of such federation needs to answer "what should be connected".

I think on the baseline, issues and comments are to be handled by the target repository server. But if the federation does not handle PRs and Forks, the federation does not make any sense.

In longer terms, issues and comments can be federated the same way as `Post` as specified in ActivityPub. We might not need to specify what is already an open standard.

**Comment by 21stio** [comment]
besides the overall vision, adding something like activitypub or gitpub will have huge implications to the underlying data model and persistence layer of gogs/gitea/gitlab.. something that can only work with the consent of the maintaners @Unknwon

**Comment by yookoala** [comment]
I'd love to have @Unknwon in the discussion / drafting of the specification.

**Comment by Unknwon** [comment]
@yookoala The idea of gitpub sounds interesting, I would love to implement if we have a open standard.

**Comment by drequivalent** [comment]
We have an ActivityPub library in Go now!

https://github.com/go-fed

**Comment by jaywink** [comment]
ForgeFed is a work in progress specification for extensions to ActivityPub for supporting forges. Would be great to get someone from Gogs involved. 

Repository: https://notabug.org/peers/forgefed/
Forum: https://talk.feneas.org/c/forgefed

**Comment by Aspie96** [comment]
This is a great idea.

Is there any progress on this issue?

**Comment by cjslep** [comment]
FYI: the github.com/go-fed/activity library has been released at `v1.0.0`, which includes complete support for the ForgeFed Vocabulary.
